### PR TITLE
upstream: changed how load calculation and panic level interact

### LINKED ISSUE #3304: rethink panic threshold and priority LB interactions
See https://github.com/envoyproxy/envoy/pull/3282. There are some ideas in the comment threads about how we can improve how these two features interact to make it more intuitive.

cc @alyssawilk

### PR DESCRIPTION
**Pull request by cpakulski**
Description: Changes for issue #3304. In a nutshell, if an LB is configured with several priority levels, panic level checking is suppressed as long as there are enough healthy hosts across all priority levels. Load distribution across levels still applies and is based on number of healthy hosts in priority set.

Risk Level: Medium

Testing: Added several unit tests to verify that load calculation is fine and panic mode is not turned if there is sufficient healthy hosts across all priority levels. 

Docs Changes: Yes - sections about load distribution and panic should be updated.

Release Notes: N/A
Fixes #3304

### CODE REVIEW

**Comment by mattklein123** [comment]
@dnoe can you take first pass on this and then I can also review?

**Comment by dnoe** [comment]
Will do.

File: source/common/upstream/load_balancer_impl.cc [hunk 1]
```diff
@@ -67,8 +68,17 @@ void LoadBalancerBase::recalculatePerPriorityState(uint32_t priority) {
                                  host_set.healthyHosts().size() / host_set.hosts().size()));
   }
 
-  // Now that we've updated health for the changed priority level, we need to caculate percentage
+  // Now that we've updated health for the changed priority level, we need to calculate percentage
   // load for all priority levels.
+
+  // The following cases are handled here;
+  // - Total health is >= 100. It means there are enough healthy hosts to handle the load.
```
[the remark this answers is not in the document]
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: Done.

File: source/common/upstream/load_balancer_impl.cc [hunk 2]
```diff
@@ -88,6 +99,11 @@ void LoadBalancerBase::recalculatePerPriorityState(uint32_t priority) {
     per_priority_load_[i] =
         std::min<uint32_t>(total_load, per_priority_health_[i] * 100 / total_health);
     total_load -= per_priority_load_[i];
+
+    // For each level check if it should run in Panic mode. Never set Panic mode if the total health
+    // is 100%, even when individual priority level has very low # of healthy hosts.
+    HostSet& priority_host_set = *priority_set_.hostSetsPerPriority()[i];
```
[the remark this answers is not in the document]
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: Done.

File: source/common/upstream/maglev_lb.h [hunk 3]
```diff
@@ -58,13 +58,13 @@ class MaglevLoadBalancer : public ThreadAwareLoadBalancerBase {
 
 private:
   // ThreadAwareLoadBalancerBase
-  HashingLoadBalancerSharedPtr createLoadBalancer(const HostSet& host_set) override {
+  HashingLoadBalancerSharedPtr createLoadBalancer(const HostSet& host_set, bool inPanic) override {
```
[the remark this answers is not in the document]
Comment (cpakulski) [reply source/common/upstream/maglev_lb.h]: Done.

File: test/common/upstream/load_balancer_impl_test.cc [hunk 4]
```diff
@@ -71,14 +72,25 @@ class LoadBalancerBaseTest : public LoadBalancerTestBase {
...
 
-  std::vector<uint32_t> getLoadPercentage() {
-    std::vector<uint32_t> ret;
+  template <typename T, typename FUNC>
+  std::vector<T> aggregatePrioritySetsValues(TestLb& lb, FUNC func) {
+    std::vector<T> ret;
+
     for (size_t i = 0; i < priority_set_.host_sets_.size(); ++i) {
-      ret.push_back(lb_.percentageLoad(i));
+      ret.push_back((lb.*func)(i));
```
[the remark this answers is not in the document]
Comment (cpakulski) [reply test/common/upstream/load_balancer_impl_test.cc]: No, it does not work. Compiler tries to call lb's func method and complains that method is not defined.

**Comment by cpakulski** [comment]
@dnoe Thanks for reviewing.

**Comment by dnoe** [comment]
@mattklein123 PTAL

File: test/common/upstream/load_balancer_impl_test.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (mattklein123) [reply test/common/upstream/load_balancer_impl_test.cc]: Perhaps just pass an std::function instead of a templatized function? That would definitely be more readable. Up to you.

File: source/common/upstream/load_balancer_impl.cc [hunk 5]
```diff
@@ -67,8 +68,17 @@ void LoadBalancerBase::recalculatePerPriorityState(uint32_t priority) {
                                  host_set.healthyHosts().size() / host_set.hosts().size()));
   }
 
-  // Now that we've updated health for the changed priority level, we need to caculate percentage
+  // Now that we've updated health for the changed priority level, we need to calculate percentage
   // load for all priority levels.
+
+  // The following cases are handled here;
```
Comment (mattklein123) [inline source/common/upstream/load_balancer_impl.cc]: I think we should update https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/load_balancing.html#panic-threshold?

File: source/common/upstream/load_balancer_impl.cc [hunk 6]
```diff
@@ -88,6 +99,11 @@ void LoadBalancerBase::recalculatePerPriorityState(uint32_t priority) {
     per_priority_load_[i] =
         std::min<uint32_t>(total_load, per_priority_health_[i] * 100 / total_health);
     total_load -= per_priority_load_[i];
+
+    // For each level check if it should run in Panic mode. Never set Panic mode if the total health
+    // is 100%, even when individual priority level has very low # of healthy hosts.
+    const HostSet& priority_host_set = *priority_set_.hostSetsPerPriority()[i];
+    per_priority_panic_[i] = (total_health == 100 ? false : isGlobalPanic(priority_host_set));
```
Comment (mattklein123) [inline source/common/upstream/load_balancer_impl.cc]: As a reader, I wonder if everywhere here we talk about "total health" we really should call it something else like "normalized total health" or something like that? It's a little confusing as we apply a min function over all priorities. Thoughts? Mostly just trying to make this very complicated code easier to understand for the next person to come along. cc @alyssawilk and @htuch also who have worked in this area.

File: test/common/upstream/load_balancer_impl_test.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (cpakulski) [reply test/common/upstream/load_balancer_impl_test.cc]: I used template because logic is the same (iterate through priority sets), but type of return value is different: `vector<uint32_t>` for percentageLoad and `vector<bool>` for panic setting.

[the remark this answers is not in the document]
Comment (mattklein123) [reply test/common/upstream/load_balancer_impl_test.cc]: I think you could templatize the return type of std::function. In any case, it's not a big deal either way.

File: source/common/upstream/load_balancer_impl.cc [hunk 7]
```diff
@@ -67,8 +68,17 @@ void LoadBalancerBase::recalculatePerPriorityState(uint32_t priority) {
...
   }
 
-  // Now that we've updated health for the changed priority level, we need to caculate percentage
+  // Now that we've updated health for the changed priority level, we need to calculate percentage
   // load for all priority levels.
+
+  // The following cases are handled here;
+  // - Total health is = 100. It means there are enough healthy hosts to handle the load.
+  //   Do not enter panic mode, even if a specific priority has low number of healthy hosts.
+  // - Total health is < 100. There is not enough healthy hosts to handle the load. Continue
```
Comment (alyssawilk) [inline source/common/upstream/load_balancer_impl.cc]: There is not -> There are not

File: source/common/upstream/load_balancer_impl.cc [hunk 8]
```diff
@@ -67,8 +68,17 @@ void LoadBalancerBase::recalculatePerPriorityState(uint32_t priority) {
...
+  //   distibuting the load among priority sets, but turn on Panic mode if # of healthy hosts
```
Comment (alyssawilk) [inline source/common/upstream/load_balancer_impl.cc]: but turn on Panic mode -> turn on panic mode for a given priority

File: source/common/upstream/load_balancer_impl.cc [hunk 9]
```diff
@@ -67,8 +68,17 @@ void LoadBalancerBase::recalculatePerPriorityState(uint32_t priority) {
...
+  //   in priority set is low.
+  // - Total health is 0. All hosts are down. Redirect 100% of traffic to P=0 and enable PanicMode.
```
Comment (alyssawilk) [inline source/common/upstream/load_balancer_impl.cc]: super nitty nit but I see panic mode, Panic mode, and PanicMode in this comment - mind picking one and updating the comments in this PR?

File: source/common/upstream/load_balancer_impl.cc [hunk 10]
```diff
@@ -88,6 +99,11 @@ void LoadBalancerBase::recalculatePerPriorityState(uint32_t priority) {
     per_priority_load_[i] =
         std::min<uint32_t>(total_load, per_priority_health_[i] * 100 / total_health);
     total_load -= per_priority_load_[i];
+
+    // For each level check if it should run in Panic mode. Never set Panic mode if the total health
+    // is 100%, even when individual priority level has very low # of healthy hosts.
```
Comment (alyssawilk) [inline source/common/upstream/load_balancer_impl.cc]: This plays interestingly with things like #4529 so I just want it on @snowp 's radar.  

For simplicity's sake I'd prefer that is if we're only temporarily excluding a priority level due to retries we not recalculate per_priority_panic_ based on if we'd turn up panic mode if that priority were truly not in play.  Realistically if we're doing retries it's probably because things we hope were healthy are not, and we if it's bad enough  we're failing over we may also want to lean in the direction of panic mode.  I'll leave that up to Snow :-P

File: source/common/upstream/load_balancer_impl.cc (hunk 6 above)
Comment (alyssawilk) [reply source/common/upstream/load_balancer_impl.cc]: Yeaaaah, I encourage making it easier to understand but I don't think I have better advice than yours :-/

File: source/common/upstream/load_balancer_impl.cc (hunk 10 above)
Comment (snowp) [reply source/common/upstream/load_balancer_impl.cc]: As #4529 stands this won't be affected by the excluded priorities since I rebalance based on those exclusions *after* calling `recalculatePerPriorityState`. We'd be using the priority load adjusted for panic mode and redistribute the load of excluded priorities between the ones not excluded, which I think is a sane behavior. You'd honor panic mode but make an extra effort to spread retries between distinct priorities.

tl;dr I think this is fine as is

**Comment by cpakulski** [comment]
@alyssawilk Thanks for reviewing the changes. I will update the code as per your suggestions. 

@mattklein123 @alyssawilk @snowp Could you voice your opinion on the following idea?  Currently `recalculatePerPriorityState` handles the following cases:
- 100 % health ->> no panic mode and redistribute traffic according to # of heathy hosts
- health < 100% ->> redistribute traffic according to # of heathy hosts in priorities and turn on panic mode if necessary
- health = 0 ->> 100% of traffic to P=0 and P=0 in panic mode.

When I was writing unit tests, one of test cases was: 3 priorities with 5 hosts each. P=0 and P=1 are completely dead (zero healthy hosts) and P=2 has only one healthy host. The current implementation will send zero traffic to P=0 and P=1 and 100% of the traffic to P=2, which is in panic mode and has only one healthy host.

The idea is that when ALL priority levels are in panic mode, the load distribution algorithm changes and calculates percentage of the total traffic based not on the number of healthy hosts, but on number of total hosts in a priority set. So in the above example, the distribution of load would be 34% (P=0), 33% (P=1) and 33% (P=2). All are in panic mode, so traffic will blast to all defined hosts in the cluster.
The point when load calculation changes from using number of healthy hosts to number of total hosts would be when ALL priority levels enter panic mode.

**Comment by mattklein123** [comment]
> The idea is that when ALL priority levels are in panic mode, the load distribution algorithm changes and calculates percentage of the total traffic based not on the number of healthy hosts, but on number of total hosts in a priority set. So in the above example, the distribution of load would be 34% (P=0), 33% (P=1) and 33% (P=2). All are in panic mode, so traffic will blast to all defined hosts in the cluster.
The point when load calculation changes from using number of healthy hosts to number of total hosts would be when ALL priority levels enter panic mode.

This makes sense to me, though I wonder if we should do this in stages? IMO what you have in this PR is a big improvement and is pretty easy to understand. If people end up wanting further refinement we can iterate? I don't feel very strongly either way though. Will defer to what @alyssawilk and @snowp think.

**Comment by snowp** [comment]
+1 on iterating on this in later PRs. This seems like a clear improvement. 

I like the suggestion, it seems to better utilize the total capacity of the cluster when panic mode is triggered, especially in the case where total health = 0 where it will distribute load between all priorities instead of just P0. If I understand it correctly, the current solution won't ever route to a 0% healthy priority, and the suggestion seems to solve that.

File: source/common/upstream/load_balancer_impl.cc (hunk 10 above)
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: I was ready for another code push but did rebase, because github complained about conflict. After @snowp changes for #4529, `recalculatePerPriorityState` is static method now. In this PR, I added call to `isGlobalPanic` which is not static and requires LoadBalancerBase object to access configuration for panic thresholds. Probably I have to calculate panic states outside of `recalculatePerPriorityState` or pass more params to it, which is bit ugly.

Comment (snowp) [reply source/common/upstream/load_balancer_impl.cc]: I think adding it as a parameter isn't that bad. If the parameter list grows too big maybe we should add something like a `PerPriorityLoadBalancerState` struct to encapsulate all the per priority state

**Comment by cpakulski** [comment]
> If I understand it correctly, the current solution won't ever route to a 0% healthy priority, and the suggestion seems to solve that.

@snowp That is correct. 

I will code it as separate PR.

File: docs/root/intro/arch_overview/load_balancing.rst [hunk 11]
```diff
@@ -247,6 +227,72 @@ To sum this up in pseudo algorithms:
   total_health = min(100, Σ(health(P_0)...health(P_X))
   load to P_X = 100 - Σ(percent_load(P_0)..percent_load(P_X-1))
 
+.. _arch_overview_load_balancing_panic_threshold:
```
Comment (mattklein123) [inline docs/root/intro/arch_overview/load_balancing.rst]: This change is definitely worthy of a release note. Can you add one?

File: docs/root/intro/arch_overview/load_balancing.rst [hunk 12]
```diff
@@ -247,6 +227,72 @@ To sum this up in pseudo algorithms:
...
+However, if the percentage of healthy hosts in the cluster becomes too low, Envoy will disregard
+health status and balance amongst all hosts. This is known as the *panic threshold*. The default
+panic threshold is 50%. This is :ref:`configurable <config_cluster_manager_cluster_runtime>` via
+runtime as well as in the :ref:`cluster configuration
+<envoy_api_field_Cluster.CommonLbConfig.healthy_panic_threshold>`. The panic threshold
+is used to avoid a situation in which host failures cascade throughout the cluster as load
+increases.
+
+Panic thresholds work in conjunction with priorities. If number of healthy hosts in given priority
+goes down, Envoy will try try shift some traffic to lower priorities. If it succeeds finding enough
```
Comment (mattklein123) [inline docs/root/intro/arch_overview/load_balancing.rst]: s/try try/try to

File: docs/root/intro/arch_overview/load_balancing.rst [hunk 13]
```diff
@@ -247,6 +227,72 @@ To sum this up in pseudo algorithms:
...
+panic threshold is 50%. This is :ref:`configurable <config_cluster_manager_cluster_runtime>` via
+runtime as well as in the :ref:`cluster configuration
+<envoy_api_field_Cluster.CommonLbConfig.healthy_panic_threshold>`. The panic threshold
+is used to avoid a situation in which host failures cascade throughout the cluster as load
+increases.
+
+Panic thresholds work in conjunction with priorities. If number of healthy hosts in given priority
+goes down, Envoy will try try shift some traffic to lower priorities. If it succeeds finding enough 
+healthy hosts in lower priorities, Envoy will disregard panic thresholds. In mathematical terms, 
+if total health across all priority levels is 100%, Envoy disregards panic thresholds but continues to
```
Comment (mattklein123) [inline docs/root/intro/arch_overview/load_balancing.rst]: Can you explain a bit more what "total health" here actually refers to? This is related to my comment in the code itself in which IMO this is not very intuitive. Perhaps actually described that it's normalized by adding up the health of each priority and capping at 100% or something like that?

File: source/common/upstream/load_balancer_impl.cc [hunk 14]
```diff
@@ -47,12 +47,29 @@ LoadBalancerBase::LoadBalancerBase(const PrioritySet& priority_set, ClusterStats
...
+  priority_set_.addMemberUpdateCb(
+      [this](uint32_t priority, const HostVector&, const HostVector&) -> void {
+        UNREFERENCED_PARAMETER(priority);
+        recalculatePerPriorityPanic();
+      });
 }
 
+// The following cases are handled by
+// recalculatePerPriorityState and recalculatePerPriorityPanic methods;
+// - Total health is = 100. It means there are enough healthy hosts to handle the load.
```
Comment (mattklein123) [inline source/common/upstream/load_balancer_impl.cc]: Can we refer to this as "normalized total health" with a description of what that means? Same below in comment?

File: source/common/upstream/load_balancer_impl.cc [hunk 15]
```diff
@@ -107,6 +127,29 @@ void LoadBalancerBase::recalculatePerPriorityState(uint32_t priority,
   }
 }
 
+// Method iterates through priority levels and turns on/off panic mode.
+void LoadBalancerBase::recalculatePerPriorityPanic() {
+  per_priority_panic_.resize(priority_set_.hostSetsPerPriority().size());
+
+  const uint32_t normalized_total_health = std::min<uint32_t>(
```
Comment (mattklein123) [inline source/common/upstream/load_balancer_impl.cc]: It seems we are doing this calculation in two places now. Can we make a shared helper with good comments?

File: docs/root/intro/arch_overview/load_balancing.rst (hunk 12 above)
Comment (htuch) [inline docs/root/intro/arch_overview/load_balancing.rst]: Nit: "succeeds in finding"

File: docs/root/intro/arch_overview/load_balancing.rst [hunk 16]
```diff
@@ -247,6 +227,72 @@ To sum this up in pseudo algorithms:
...
+During load balancing, Envoy will generally only consider healthy hosts in an upstream cluster.
+However, if the percentage of healthy hosts in the cluster becomes too low, Envoy will disregard
+health status and balance amongst all hosts. This is known as the *panic threshold*. The default
+panic threshold is 50%. This is :ref:`configurable <config_cluster_manager_cluster_runtime>` via
+runtime as well as in the :ref:`cluster configuration
+<envoy_api_field_Cluster.CommonLbConfig.healthy_panic_threshold>`. The panic threshold
+is used to avoid a situation in which host failures cascade throughout the cluster as load
+increases.
+
+Panic thresholds work in conjunction with priorities. If number of healthy hosts in given priority
```
Comment (htuch) [inline docs/root/intro/arch_overview/load_balancing.rst]: Nit: `in a given priority`

File: docs/root/intro/arch_overview/load_balancing.rst [hunk 17]
```diff
@@ -247,6 +227,72 @@ To sum this up in pseudo algorithms:
...
++------------------------+-------------------------+-----------------+-----------------+-----------------+
+| 71%                    |  71%                    | 100%            |   NO            | NO              |
++------------------------+-------------------------+-----------------+-----------------+-----------------+
+| 50%                    |  50%                    | 100%            |   NO            | NO              |
++------------------------+-------------------------+-----------------+-----------------+-----------------+
+| 25%                    |  100%                   | 100%            |   NO            | NO              |
++------------------------+-------------------------+-----------------+-----------------+-----------------+
+| 25%                    |  25%                    | 70%             |   YES           | YES             |
++------------------------+-------------------------+-----------------+-----------------+-----------------+
+| 5%                     |  65%                    | 98%             |   YES           | NO              |
```
Comment (htuch) [inline docs/root/intro/arch_overview/load_balancing.rst]: I wasn't in the original discussion, but can you TL;DR summarize what the thinking was in allowing P=0 to enter panic (and hence spray to unhealthy host) while P=1 seems happier to sink the traffic in this case?

File: docs/root/intro/arch_overview/load_balancing.rst [hunk 18]
```diff
@@ -247,6 +227,72 @@ To sum this up in pseudo algorithms:
...
++----------------------------+--------------------+
+| 0%                         | NO                 |
++----------------------------+--------------------+
+
+If P=1 becomes unhealthy, panic threshold continues to be disregarded until the sum of the health
+P=0 + P=1 goes below 100. At this point Envoy starts checking panic threshold value for each 
+priority.
+
++------------------------+-------------------------+-----------------+-----------------+-----------------+
+| P=0 healthy endpoints  | P=1 healthy endpoints   | Total health    |  P=0 in panic   | P=1 in panic    |
```
Comment (htuch) [inline docs/root/intro/arch_overview/load_balancing.rst]: Should there be columns for the effective amount of traffic delivered at P=0 and P=1?

File: docs/root/intro/arch_overview/load_balancing.rst (hunk 17 above)
Comment (htuch) [reply docs/root/intro/arch_overview/load_balancing.rst]: (and maybe try and convey this intuition regarding it being the sensible thing to do in the writeup here)

File: docs/root/intro/arch_overview/load_balancing.rst (hunk 11 above)
Comment (cpakulski) [reply docs/root/intro/arch_overview/load_balancing.rst]: Sure, where is file with release notes?

File: docs/root/intro/arch_overview/load_balancing.rst (hunk 12 above)
Comment (cpakulski) [reply docs/root/intro/arch_overview/load_balancing.rst]: Done.

File: docs/root/intro/arch_overview/load_balancing.rst (hunk 13 above)
Comment (cpakulski) [reply docs/root/intro/arch_overview/load_balancing.rst]: Done.

File: source/common/upstream/load_balancer_impl.cc (hunk 14 above)
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: Done.

File: source/common/upstream/load_balancer_impl.cc (hunk 15 above)
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: Done. Added new method to calculate normalized total health.

File: docs/root/intro/arch_overview/load_balancing.rst (hunk 12 above)
Comment (cpakulski) [reply docs/root/intro/arch_overview/load_balancing.rst]: Done.

File: docs/root/intro/arch_overview/load_balancing.rst (hunk 16 above)
Comment (cpakulski) [reply docs/root/intro/arch_overview/load_balancing.rst]: Done.

File: docs/root/intro/arch_overview/load_balancing.rst (hunk 17 above)
Comment (cpakulski) [reply docs/root/intro/arch_overview/load_balancing.rst]: Tried my best.

File: docs/root/intro/arch_overview/load_balancing.rst (hunk 18 above)
Comment (cpakulski) [reply docs/root/intro/arch_overview/load_balancing.rst]: Added.

File: docs/root/intro/arch_overview/load_balancing.rst (hunk 11 above)
Comment (mattklein123) [reply docs/root/intro/arch_overview/load_balancing.rst]: https://github.com/envoyproxy/envoy/blob/master/docs/root/intro/version_history.rst

File: source/common/upstream/load_balancer_impl.cc [hunk 19]
```diff
@@ -107,6 +134,29 @@ void LoadBalancerBase::recalculatePerPriorityState(uint32_t priority,
...
 
+// Method iterates through priority levels and turns on/off panic mode.
+void LoadBalancerBase::recalculatePerPriorityPanic() {
+  per_priority_panic_.resize(priority_set_.hostSetsPerPriority().size());
+
+  const uint32_t normalized_total_health = calcNormalizedTotalHealth(per_priority_health_);
+
+  if (normalized_total_health == 0) {
+    // Everything is terrible. All load should be to P=0. Turn on panic mode.
+    ASSERT(per_priority_load_[0] == 100);
```
[the remark this answers is not in the document]
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: > Question - I know we had behavior like this before, but is this "everything to P0 in panic" exactly what we want? What I'm thinking about is a situation where we have 0% healthy in P0 and 1% healthy at P1. At this point, P0 and P1 are in panic, total normalized health is 1%. I think P1 should be basically getting all the traffic and spraying across all hosts. Now, P1 becomes 0% healthy. Suddenly, all load swings to P0. Then, when some P1 hosts become healthy, it swings the other way.

That is exactly how it works now. I can definitely add a section describing behaviour when the number of healthy hosts is really low. However, for exactly such scenario  I suggested the following earlier in this PR:

> The idea is that when ALL priority levels are in panic mode, the load distribution algorithm changes and calculates percentage of the total traffic based not on the number of healthy hosts, but on number of total hosts in a priority set. So in the above example, the distribution of load would be 34% (P=0), 33% (P=1) and 33% (P=2). All are in panic mode, so traffic will blast to all defined hosts in the cluster.
The point when load calculation changes from using number of healthy hosts to number of total hosts would be when ALL priority levels enter panic mode.

Based on feedback from few people, it seems like idea worth considering. WDYT?
In your example P0 (0% healthy) and P1(1% healthy) traffic would be distributed 50/50 (assuming P0 and P1 have the same number of total hosts). It brings some stability and avoids flapping and moving 100% of traffic to barely healthy priority. Sure, traffic is sprayed across totally unhealthy priority levels, but what else can be done? With such low number of healthy hosts, we can safely assume that most of requests would not be fulfilled anyways if 100% of traffic goes to P1 with 1% of healthy hosts. Actually, moving 100% of traffic to P1, may overload the remaining healthy hosts and again traffic would go back to P=0.
I would code it as separate PR.

[the remark this answers is not in the document]
Comment (htuch) [reply source/common/upstream/load_balancer_impl.cc]: This seems reasonable. Can you open an issue to track this? @alyssawilk @mattklein123 WDYT?

[the remark this answers is not in the document]
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: Done. #4685. You can assign it to me. Thanks.

[the remark this answers is not in the document]
Comment (mattklein123) [reply source/common/upstream/load_balancer_impl.cc]: +1, thanks

File: source/common/upstream/load_balancer_impl.cc [hunk 20]
```diff
@@ -47,12 +47,32 @@ LoadBalancerBase::LoadBalancerBase(const PrioritySet& priority_set, ClusterStats
...
+      });
 }
 
+// The following cases are handled by
+// recalculatePerPriorityState and recalculatePerPriorityPanic methods (normalized total health is
+// sum of all priorities' health values and capped at 100).
+// - normalized total health is = 100%. It means there are enough healthy hosts to handle the load.
+//   Do not enter panic mode, even if a specific priority has low number of healthy hosts.
+// - normalized total health is < 100%. There are not enough healthy hosts to handle the load.
+// Continue
```
Comment (mattklein123) [inline source/common/upstream/load_balancer_impl.cc]: nit: reflow comment. Feel free to do this in your next PR.
